helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java

```diff
@@ -643,14 +643,17 @@ public void init() {
     logger.info("initializing CallbackHandler: {}, content: {} ", _uid, getContent());
 
     if (_batchModeEnabled) {
-      synchronized (this) {
-        if (_batchCallbackProcessor != null) {
-          _batchCallbackProcessor.resetEventQueue();
+      CallbackProcessor callbackProcessorRef = _batchCallbackProcessor.get();
+        if (callbackProcessorRef != null) {
+          callbackProcessorRef.resetEventQueue();
         } else {
-          _batchCallbackProcessor = new CallbackProcessor(this);
-          _batchCallbackProcessor.start();
+          callbackProcessorRef = new CallbackProcessor(this);
+          if (_batchCallbackProcessor.compareAndSet(null, callbackProcessorRef)) {
+            callbackProcessorRef.start();
```

Comment:
Consider the following operation order:
1. init() and reset() is called at the same time.
2. In init(), the code reaches line 652, but not start yet.
3. At the same time, in reset(), code reaches line 776, and tries to shutdown the thread. But it cannot, since the thread has not been started yet. Note, in java doc, "Interrupting a thread that is not alive need not have any effect."
4. Then the code in init() try to start, the thread is started.

We have a leaked thread, eventually.

Reply:
I think you will need to start and then insert it into the atomic object.

But please revisit and see if there are any other ways.

Reply:
TFTR. I think you are correct. We should't touch the `callbackProcessorRef ` object after `compareAndSet` returns true.
Need to start and then insert